File: services/company/operations/services/onboard-dimension-fields.ts
import { APIError } from "encore.dev/api";

// Bảng field hợp lệ cho từng chiều onboarding — nguồn sự thật duy nhất cho payload
// `POST /operations/onboard/dimensions/:dimension`. apps/cosa giữ bảng tương ứng ở
// `apps/cosa/workflows/onboarding_dimensions.py`; tests/contracts/
// test_startup_os_dimension_fields.py đối chiếu hai bên (lệch ⇒ dữ liệu agent gửi
// bị bỏ âm thầm, đúng lỗi đã xảy ra khi workflow gửi `arr`/`burn_rate_weekly`).
//
// Kiểu field:
// - string / integer / number / boolean: giá trị đơn, cho phép null.
// - string_array: mảng chuỗi.
// - value_list: mảng chuỗi hoặc { valueText, isFireWorthy?, realityStatus? }.
// - competitor_list: mảng { name, whyWinning?, threatLevel? }.
export type OnboardFieldType =
  | "string"
  | "integer"
  | "number"
  | "boolean"
  | "string_array"
  | "value_list"
  | "competitor_list";

export const ONBOARD_DIMENSION_FIELDS = {
  identity: {
    whatTheyDo: "string",
    whoTheyServe: "string",
    foundingWhy: "string",
    oneSentencePitch: "string",
    values: "value_list",
    notCaptured: "string_array",
  },
  stage_scale: {
    headcountFt: "integer",
    headcountContractor: "integer",
    revenueArr: "number",
    revenueCurrency: "string",
    runwayMonths: "number",
    stage: "string",
    whatBrokeLast90d: "string",
    notCaptured: "string_array",
  },
  founder: {
    founderName: "string",
    role: "string",
    superpower: "string",
    blindSpots: "string",
    archetype: "string",
    whatKeepsUp: "string",
    cofounderCritique: "string",
    notCaptured: "string_array",
  },
  team_culture: {
    threeWords: "string_array",
    lastRealConflict: "string",
    conflictResolution: "string",
    strongestLeader: "string",
    weakestLeader: "string",
    hasRealConflict: "boolean",
    notCaptured: "string_array",
  },
  market: {
    marketDescription: "string",
    unfairAdvantage: "string",
    competitiveThreat: "string",
    hasRealCompetition: "boolean",
    competitors: "competitor_list",
    notCaptured: "string_array",
  },
  challenges: {
    priorityProduct: "integer",
    priorityGrowth: "integer",
    priorityPeople: "integer",
    priorityMoney: "integer",
    priorityOperations: "integer",
    avoidedDecision: "string",
    extraDayAnswer: "string",
    notCaptured: "string_array",
  },
  goals_ambition: {
    goal12MonthsText: "string",
    goal36MonthsText: "string",
    exitOrientation: "string",
    personalSuccessDefinition: "string",
    notCaptured: "string_array",
  },
} as const satisfies Record<string, Record<string, OnboardFieldType>>;

export type OnboardDimension = keyof typeof ONBOARD_DIMENSION_FIELDS;

// Giá trị enum lưu dạng text trong DB (xem comment schema onboard.ts).
const ENUM_VALUES: Partial<Record<string, readonly string[]>> = {
  stage: ["pre_pmf", "scaling", "optimizing"],
  archetype: ["product", "sales", "technical", "operator", "hybrid"],
  exitOrientation: ["exit", "build_forever", "undecided"],
};

const PRIORITY_FIELDS = new Set([
  "priorityProduct",
  "priorityGrowth",
  "priorityPeople",
  "priorityMoney",
  "priorityOperations",
]);

export function isOnboardDimension(value: string): value is OnboardDimension {
  return Object.prototype.hasOwnProperty.call(ONBOARD_DIMENSION_FIELDS, value);
}

function fail(dimension: string, field: string, reason: string): never {
  throw APIError.invalidArgument(`onboard dimension '${dimension}' field '${field}': ${reason}`);
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function checkField(dimension: string, field: string, type: OnboardFieldType, value: unknown): void {
  if (value === null || value === undefined) return;
  switch (type) {
    case "string":
      if (typeof value !== "string") fail(dimension, field, "must be a string");
      {
        const allowed = ENUM_VALUES[field];
        if (allowed && !allowed.includes(value)) {
          fail(dimension, field, `must be one of ${allowed.join(", ")}`);
        }
      }
      return;
    case "integer":
      if (typeof value !== "number" || !Number.isInteger(value)) fail(dimension, field, "must be an integer");
      if (PRIORITY_FIELDS.has(field) && (value < 1 || value > 5)) fail(dimension, field, "must be between 1 and 5");
      if (value < 0) fail(dimension, field, "must not be negative");
      return;
    case "number":
      if (typeof value !== "number" || !Number.isFinite(value)) fail(dimension, field, "must be a finite number");
      if (value < 0) fail(dimension, field, "must not be negative");
      return;
    case "boolean":
      if (typeof value !== "boolean") fail(dimension, field, "must be a boolean");
      return;
    case "string_array":
      if (!Array.isArray(value) || value.some((v) => typeof v !== "string")) {
        fail(dimension, field, "must be an array of strings");
      }
      return;
    case "value_list":
      if (!Array.isArray(value)) fail(dimension, field, "must be an array");
      for (const item of value) {
        if (typeof item === "string") continue;
        if (!isPlainObject(item) || typeof item.valueText !== "string" || item.valueText.trim() === "") {
          fail(dimension, field, "each item must be a string or { valueText }");
        }
        if (item.isFireWorthy !== undefined && typeof item.isFireWorthy !== "boolean") {
          fail(dimension, field, "isFireWorthy must be a boolean");
        }
        if (
          item.realityStatus !== undefined &&
          !["real", "poster", "unclear"].includes(String(item.realityStatus))
        ) {
          fail(dimension, field, "realityStatus must be one of real, poster, unclear");
        }
      }
      return;
    case "competitor_list":
      if (!Array.isArray(value)) fail(dimension, field, "must be an array");
      for (const item of value) {
        if (!isPlainObject(item) || typeof item.name !== "string" || item.name.trim() === "") {
          fail(dimension, field, "each competitor must be { name }");
        }
        if (item.whyWinning !== undefined && typeof item.whyWinning !== "string") {
          fail(dimension, field, "whyWinning must be a string");
        }
        if (
          item.threatLevel !== undefined &&
          !["low", "medium", "high"].includes(String(item.threatLevel))
        ) {
          fail(dimension, field, "threatLevel must be one of low, medium, high");
        }
      }
      return;
  }
}

/**
 * Kiểm payload 1 chiều trước khi ghi. Field lạ bị từ chối (không bỏ âm thầm) để
 * lệch hợp đồng giữa client/agent và Company lộ ra ngay tại request.
 */
export function validateOnboardDimensionData(
  dimension: string,
  data: unknown
): { dimension: OnboardDimension; data: Record<string, unknown> } {
  if (!isOnboardDimension(dimension)) {
    throw APIError.invalidArgument(`Chiều onboarding không hợp lệ: ${dimension}`);
  }
  if (!isPlainObject(data)) {
    throw APIError.invalidArgument(`onboard dimension '${dimension}': data must be an object`);
  }
  const fields: Record<string, OnboardFieldType> = ONBOARD_DIMENSION_FIELDS[dimension];
  const unknown = Object.keys(data).filter((key) => !(key in fields));
  if (unknown.length > 0) {
    throw APIError.invalidArgument(
      `onboard dimension '${dimension}': unknown field(s) ${unknown.join(", ")}; ` +
        `allowed: ${Object.keys(fields).join(", ")}`
    );
  }
  for (const [field, type] of Object.entries(fields)) {
    checkField(dimension, field, type, data[field]);
  }
  const captured = Object.keys(fields).filter(
    (field) => field !== "notCaptured" && data[field] !== undefined && data[field] !== null
  );
  if (captured.length === 0) {
    throw APIError.invalidArgument(`onboard dimension '${dimension}': at least one field must be provided`);
  }
  if (dimension === "founder" && (typeof data.founderName !== "string" || data.founderName.trim() === "")) {
    fail(dimension, "founderName", "is required");
  }
  return { dimension, data };
}

File: frontend/lib/modules/startup_os/models/onboard_dimension_fields.dart
// Field của 7 chiều onboarding cho wizard form (plan 2026-09-18 Task 4.3).
//
// Key và kiểu PHẢI khớp `services/company/operations/services/onboard-dimension-fields.ts`
// — Company từ chối field lạ. test/modules/startup_os/onboard_dimension_fields_test.dart
// đối chiếu trực tiếp với file TS. `notCaptured` không có trên form: wizard tự
// điền tên các field Founder để trống.

enum OnboardFieldKind {
  string,
  integer,
  number,
  boolean,
  stringArray,
  valueList,
  competitorList,
}

extension OnboardFieldKindWire on OnboardFieldKind {
  String get wire => switch (this) {
        OnboardFieldKind.string => 'string',
        OnboardFieldKind.integer => 'integer',
        OnboardFieldKind.number => 'number',
        OnboardFieldKind.boolean => 'boolean',
        OnboardFieldKind.stringArray => 'string_array',
        OnboardFieldKind.valueList => 'value_list',
        OnboardFieldKind.competitorList => 'competitor_list',
      };
}

class OnboardField {
  const OnboardField(this.key, this.label, this.kind, {this.options, this.hint});

  final String key;
  final String label;
  final OnboardFieldKind kind;

  /// Giá trị enum (dropdown) — key giá trị wire → nhãn hiển thị.
  final Map<String, String>? options;
  final String? hint;
}

class OnboardDimensionForm {
  const OnboardDimensionForm({
    required this.dimension,
    required this.title,
    required this.isFast,
    required this.fields,
  });

  final String dimension;
  final String title;
  final bool isFast;
  final List<OnboardField> fields;
}

const _priorityOptions = {
  '1': '1 — thấp',
  '2': '2',
  '3': '3',
  '4': '4',
  '5': '5 — cao nhất',
};

/// Thứ tự giống kịch bản hội thoại /cs:setup (Fast → Medium → Slow).
const List<OnboardDimensionForm> onboardDimensionForms = [
  OnboardDimensionForm(
    dimension: 'stage_scale',
    title: 'Giai đoạn & Quy mô',
    isFast: true,
    fields: [
      OnboardField('stage', 'Giai đoạn', OnboardFieldKind.string, options: {
        'pre_pmf': 'Chưa đạt PMF',
        'scaling': 'Đang scale',
        'optimizing': 'Tối ưu vận hành',
      }),
      OnboardField('revenueArr', 'Doanh thu năm quy đổi (ARR)', OnboardFieldKind.number),
      OnboardField('revenueCurrency', 'Đơn vị tiền tệ', OnboardFieldKind.string, hint: 'VD: VND, USD'),
      OnboardField('runwayMonths', 'Runway (tháng)', OnboardFieldKind.number),
      OnboardField('headcountFt', 'Nhân sự toàn thời gian', OnboardFieldKind.integer),
      OnboardField('headcountContractor', 'Cộng tác viên / hợp đồng', OnboardFieldKind.integer),
      OnboardField('whatBrokeLast90d', 'Điều gì đã trục trặc trong 90 ngày qua?', OnboardFieldKind.string),
    ],
  ),
  OnboardDimensionForm(
    dimension: 'challenges',
    title: 'Thách thức & Quyết định khó',
    isFast: true,
    fields: [
      OnboardField('priorityProduct', 'Ưu tiên: Sản phẩm', OnboardFieldKind.integer, options: _priorityOptions),
      OnboardField('priorityGrowth', 'Ưu tiên: Tăng trưởng', OnboardFieldKind.integer, options: _priorityOptions),
      OnboardField('priorityPeople', 'Ưu tiên: Con người', OnboardFieldKind.integer, options: _priorityOptions),
      OnboardField('priorityMoney', 'Ưu tiên: Tiền', OnboardFieldKind.integer, options: _priorityOptions),
      OnboardField('priorityOperations', 'Ưu tiên: Vận hành', OnboardFieldKind.integer, options: _priorityOptions),
      OnboardField('avoidedDecision', 'Quyết định khó đang trì hoãn', OnboardFieldKind.string),
      OnboardField('extraDayAnswer', 'Có thêm 1 ngày/tuần sẽ làm gì?', OnboardFieldKind.string),
    ],
  ),
  OnboardDimensionForm(
    dimension: 'goals_ambition',
    title: 'Mục tiêu & Tham vọng',
    isFast: false,
    fields: [
      OnboardField('goal12MonthsText', 'Mục tiêu 12 tháng', OnboardFieldKind.string),
      OnboardField('goal36MonthsText', 'Hình dung sau 36 tháng', OnboardFieldKind.string),
      OnboardField('exitOrientation', 'Định hướng dài hạn', OnboardFieldKind.string, options: {
        'exit': 'Hướng tới exit',
        'build_forever': 'Xây dựng lâu dài',
        'undecided': 'Chưa quyết định',
      }),
      OnboardField('personalSuccessDefinition', 'Thành công với cá nhân Founder là gì?', OnboardFieldKind.string),
    ],
  ),
  OnboardDimensionForm(
    dimension: 'market',
    title: 'Thị trường & Cạnh tranh',
    isFast: false,
    fields: [
      OnboardField('marketDescription', 'Mô tả thị trường & khách hàng', OnboardFieldKind.string),
      OnboardField('unfairAdvantage', 'Lợi thế không công bằng', OnboardFieldKind.string),
      OnboardField('competitiveThreat', 'Mối đe doạ cạnh tranh lớn nhất', OnboardFieldKind.string),
      OnboardField('hasRealCompetition', 'Có đối thủ cạnh tranh trực tiếp?', OnboardFieldKind.boolean),
      OnboardField('competitors', 'Đối thủ chính', OnboardFieldKind.competitorList,
          hint: 'Mỗi dòng: Tên — vì sao họ đang thắng'),
    ],
  ),
  OnboardDimensionForm(
    dimension: 'team_culture',
    title: 'Đội ngũ & Văn hoá',
    isFast: false,
    fields: [
      OnboardField('threeWords', 'Ba từ mô tả văn hoá', OnboardFieldKind.stringArray, hint: 'Cách nhau bằng dấu phẩy'),
      OnboardField('lastRealConflict', 'Xung đột thật gần nhất', OnboardFieldKind.string),
      OnboardField('conflictResolution', 'Đã giải quyết thế nào?', OnboardFieldKind.string),
      OnboardField('hasRealConflict', 'Đội có tranh luận thẳng thắn?', OnboardFieldKind.boolean),
      OnboardField('strongestLeader', 'Người dẫn dắt mạnh nhất', OnboardFieldKind.string),
      OnboardField('weakestLeader', 'Vị trí lãnh đạo đang yếu nhất', OnboardFieldKind.string),
    ],
  ),
  OnboardDimensionForm(
    dimension: 'identity',
    title: 'Căn tính & Giá trị',
    isFast: false,
    fields: [
      OnboardField('whatTheyDo', 'Startup làm gì?', OnboardFieldKind.string),
      OnboardField('whoTheyServe', 'Phục vụ ai?', OnboardFieldKind.string),
      OnboardField('foundingWhy', 'Vì sao bắt đầu?', OnboardFieldKind.string),
      OnboardField('oneSentencePitch', 'Pitch một câu', OnboardFieldKind.string),
      OnboardField('values', 'Giá trị cốt lõi', OnboardFieldKind.valueList,
          hint: 'Mỗi dòng một giá trị; thêm "!" ở đầu nếu vi phạm là lý do sa thải'),
    ],
  ),
  OnboardDimensionForm(
    dimension: 'founder',
    title: 'Hồ sơ Founder',
    isFast: false,
    fields: [
      OnboardField('founderName', 'Tên Founder (bắt buộc)', OnboardFieldKind.string),
      OnboardField('role', 'Vai trò', OnboardFieldKind.string),
      OnboardField('superpower', 'Thế mạnh vượt trội', OnboardFieldKind.string),
      OnboardField('blindSpots', 'Điểm mù tự nhận', OnboardFieldKind.string),
      OnboardField('archetype', 'Kiểu Founder', OnboardFieldKind.string, options: {
        'product': 'Sản phẩm',
        'sales': 'Bán hàng',
        'technical': 'Kỹ thuật',
        'operator': 'Vận hành',
        'hybrid': 'Lai',
      }),
      OnboardField('whatKeepsUp', 'Điều gì khiến mất ngủ?', OnboardFieldKind.string),
      OnboardField('cofounderCritique', 'Co-founder hay phê bình điều gì?', OnboardFieldKind.string),
    ],
  ),
];

String dimensionTitle(String dimension) => onboardDimensionForms
    .firstWhere(
      (f) => f.dimension == dimension,
      orElse: () => OnboardDimensionForm(dimension: dimension, title: dimension, isFast: false, fields: const []),
    )
    .title;

/// Chuyển giá trị form (chuỗi người dùng nhập) thành payload đúng kiểu Company.
///
/// Trả về null khi không có field nào được điền. Field để trống được liệt kê
/// trong `notCaptured` thay vì gửi giá trị rỗng/giả. Ném [FormatException] khi
/// giá trị số không hợp lệ để form báo lỗi tại chỗ.
Map<String, Object>? buildDimensionPayload(OnboardDimensionForm form, Map<String, String> raw) {
  final data = <String, Object>{};
  final notCaptured = <String>[];

  for (final field in form.fields) {
    final text = (raw[field.key] ?? '').trim();
    if (text.isEmpty) {
      notCaptured.add(field.key);
      continue;
    }
    switch (field.kind) {
      case OnboardFieldKind.string:
        data[field.key] = text;
      case OnboardFieldKind.integer:
        final value = int.tryParse(text);
        if (value == null || value < 0) {
          throw FormatException('${field.label}: cần số nguyên không âm');
        }
        data[field.key] = value;
      case OnboardFieldKind.number:
        final value = double.tryParse(text.replaceAll(',', '.'));
        if (value == null || value < 0 || !value.isFinite) {
          throw FormatException('${field.label}: cần số không âm');
        }
        data[field.key] = value;
      case OnboardFieldKind.boolean:
        data[field.key] = text == 'true';
      case OnboardFieldKind.stringArray:
        data[field.key] =
            text.split(',').map((s) => s.trim()).where((s) => s.isNotEmpty).toList(growable: false);
      case OnboardFieldKind.valueList:
        data[field.key] = text
            .split('\n')
            .map((line) => line.trim())
            .where((line) => line.isNotEmpty)
            .map((line) => line.startsWith('!')
                ? {'valueText': line.substring(1).trim(), 'isFireWorthy': true}
                : {'valueText': line, 'isFireWorthy': false})
            .where((v) => (v['valueText'] as String).isNotEmpty)
            .toList(growable: false);
      case OnboardFieldKind.competitorList:
        data[field.key] = text
            .split('\n')
            .map((line) => line.trim())
            .where((line) => line.isNotEmpty)
            .map((line) {
              final parts = line.split(RegExp(r'\s+[—-]\s+'));
              final name = parts.first.trim();
              final why = parts.length > 1 ? parts.sublist(1).join(' - ').trim() : '';
              return <String, String>{'name': name, if (why.isNotEmpty) 'whyWinning': why};
            })
            .toList(growable: false);
    }
  }

  if (data.isEmpty) return null;
  if (form.dimension == 'founder' && !data.containsKey('founderName')) {
    throw const FormatException('Tên Founder là bắt buộc');
  }
  if (notCaptured.isNotEmpty) data['notCaptured'] = notCaptured;
  return data;
}
